CSV import for events, participants, and event participants

pjsier opened the pull request:
Unfinished PR for #23. I created the `FileImportMixin` to handle most of the CSV and data processing for views, and some basic initial UI pieces for it. After going through the code, I actually don't think much of the API code can be reused here, but a second opinion on that would be great!

I ended up having a good amount of things I wanted to flag, so including those here:

* In `change_list.html`, I'm not sure where the best place on the admin classes would be to indicate whether they have import paths, and then include the import path if they do. Figured I'd use this placeholder for now
* I was using `bulk_create` for making new objects, but in Django 1.8 it doesn't return the primary keys of the new objects, which is a bit of a dealbreaker. Leaving some of that code in there to see if there's another way (the only thing that comes to mind is checking for the most recent n objects, but that seems hack-ish)
* For `EventParticipantsImport`, I spent a while trying to think of a way to keep the many-to-many field in the general `FileImportMixin` setup, and in the end it made more sense to just override the `import_data` function.
* I was using `parse_time` to get the event time, but `TwelveHourTimeField` seems to have different rules, so not sure the best way to handle that
* On the sign-in sheet for events with the event participants import, I have the rows refreshing on import. I'm guessing this makes sense for the generate change list pages, but not sure the best way of doing that yet

cecilia-donnelly commented on streetcrm/views.py:
```diff
@@ -679,3 +716,77 @@ def get_context_data(self, **context):
+            # Use field queries (with default of the key text) to check for
+            # record existence
+            participant_query = models.Participant.objects.filter(
+                **{self.field_queries.get(key, key): value for key, value in row.items()}
```
Note to self: come back to this.

cecilia-donnelly commented on streetcrm/views.py:
```diff
@@ -679,3 +716,77 @@ def get_context_data(self, **context):
+        new_participants = models.Participant.objects.bulk_create(participants_to_create)
+        participant_ids_to_add = (
+            [p.id for p in participants_to_add] +
+            [np.id for np in new_participants]
```
Ah, so this is the problematic part, in that we don't actually know the IDs of the newly created participants?

cecilia-donnelly commented on streetcrm/static/js/import_change_list.js:
```diff
@@ -0,0 +1,26 @@
+            contentType: false,
+            data: form,
+            success: function (data) {
+                // TODO: Modal popup indicating results, potentially refresh page?
```
I like the idea of a popup telling the user "successfully imported n participants" or something like that.

cecilia-donnelly commented on streetcrm/templates/admin/change_list.html:
```diff
@@ -66,6 +67,23 @@
+                    {% endif %}>
+            </label>
+        </li>
+        {% endif %}
```
Note to self: look into this.

cecilia-donnelly replied:
@pjsier what about setting as an `attr` in `change_list_import.js`?

cecilia-donnelly commented on streetcrm/views.py:
```diff
@@ -387,6 +391,26 @@ def process_row(self, row):
+                objects_to_create.append(row)
+
+        # Create all new objects
+        # FIXME: Would bulk create, but Django 1.8 doesn't return PKs
```
Does this mean we should be upgrading to a more recent Django?  I think there were reasons we didn't (probably some breaking changes?), but at some point we'll need to.  It looks like we're 3 minor versions behind now.

pjsier replied:
I think it'd be a good idea to upgrade, and 1.11 is the most recent version and LTS, but I also don't know if there are any deprecated features we're using in the app

cecilia-donnelly commented:
Responding to @pjsier's PR message:

>In change_list.html, I'm not sure where the best place on the admin classes would be to indicate whether they have import paths, and then include the import path if they do. Figured I'd use this placeholder for now

I think that import button placement is good.  I'd somewhat prefer having it right next to the "Create new <object>" button, instead of below it.

>I was using bulk_create for making new objects, but in Django 1.8 it doesn't return the primary keys of the new objects, which is a bit of a dealbreaker. Leaving some of that code in there to see if there's another way (the only thing that comes to mind is checking for the most recent n objects, but that seems hack-ish)

Agreed that checking the most recent `n` objects is bad!  Bummer about Django 1.8.  Do more recent versions return the IDs after bulk_create?  Do you know how hard it would be to upgrade?

>For EventParticipantsImport, I spent a while trying to think of a way to keep the many-to-many field in the general FileImportMixin setup, and in the end it made more sense to just override the import_data function.

*nod*

>I was using parse_time to get the event time, but TwelveHourTimeField seems to have different rules, so not sure the best way to handle that

Hmm, neither do I.

>On the sign-in sheet for events with the event participants import, I have the rows refreshing on import. I'm guessing this makes sense for the generate change list pages, but not sure the best way of doing that yet

I'd like this, for sure, though this WITH a message would be even better.

pjsier commented:
@cecilia-donnelly the most recent commit reloads the change list pages after the notification on import and handles the `TwelveHourTimeField` issue from earlier. Going through everything mentioned here, I think that covers everything other than the `bulk_create` issue that's dependent on a Django upgrade.

I'm going to take `WIP` off of the PR because it could work as is, but do you think we should merge this without an upgrade to Django?

pjsier commented:
Good point on the toast message location! It was intended to keep it out of the way, but that was probably too out of the way to be helpful. I centered it now, and also made it wider. The increased width was mostly to accommodate the slightly more descriptive message on the sign-up import (now says how many participants were created vs. just added to the event). The message is set to display for a little longer now as well, but I added the function to remove it to a click handler so that if it's annoying someone they can click to dismiss it

For the duplicate actions, definitely let me know if that happens again, but I haven't been able to recreate it

cecilia-donnelly commented:
More review!

To test the import into an empty database, I dropped and recreated my local PostgreSQL db like so:

```
$ createdb --owner=streetcrm streetcrm
$ python manage.py migrate auth
$ python manage.py migrate
$ python manage.py createsuperuser
```

I logged back in to the app (which was now empty of data), and imported the three participants from `sample_participant_import.csv`.  That went great.  I wanted to know whether their institutions were created correctly, so I navigated in the app to the Institution change list, at http://localhost:8000/streetcrm/institution/.  There, I got an error: `AttributeError at /streetcrm/institution/ 'InstitutionAdmin' object has no attribute 'import_path'`.  Presumably this has to do with the new context you added for the import button.  Can you look into it?  I'll test to see whether this shows up on a non-empty database, too.

pjsier commented:
@cecilia-donnelly thanks for catching that! That was a bug on my part where `import_path` is loaded into the view context from `self.import_path`, but there's no default provided if it doesn't exist. Using `getattr` with a default of `None` to resolve that

cecilia-donnelly commented:
Happily, other change lists seem to be working fine and we can still look at individual institution change forms.  On that note: I should have caught this in the code but the "Import Attendees" button shows up on the institution contact list:

![import-button](https://user-images.githubusercontent.com/1497818/31901277-93ecc8ac-b7e6-11e7-8959-0ef42b7c8b72.png)

In `streetcrm/templates/admin/change_list.html`, we need a conditional that checks whether we're on an action change form or an institution change form.

pjsier commented:
Added that conditional check to the `inline_models.html` template based off of the `opts.model_name` and that seems to fix it

cecilia-donnelly commented:
Not sure how important this is, but if I click "import" while the toast message is still open then the next message doesn't come up.  So, for example, if I click "Import Participants," choose the `sample-participant-import.csv`, see the toast message that I imported zero participants (because I had already imported them), then immediately click "Import Participants" again (while the toast is still up) and choose `sample-event-import.csv` then I __don't__ get an error message.  Is that simple to fix?

pjsier commented:
@cecilia-donnelly hmm, so this is doable by removing the div with ID `toast` at the start of the `toastMessage` function, but the problem is that won't remove the `setTimeout` event, so within a second or so the toast message will still be removed by the initial timer. I imagine there's a better way of handling the timer, but I'm not sure what it is offhand

pjsier commented:
Based off of the updated issues, the most recent commit should get everything we're planning for this PR:

* Adds some additional null checks (converting blank strings to `None` in queries) on imports to prevent duplicates
* Allows the CSV import to accept previously exported files while skipping the metadata rows
* Updates the sample CSV files, adds a help page so users can find and download them 
* Adds the help page link to the import error message

I've done a good amount of testing to verify that re-importing files doesn't create duplicates, but I think that's the part I can see the most issues coming from

pjsier commented:
@cecilia-donnelly Updated this so that it can handle accepting IDs included in an import after #333, as well as ignore fields not specified in `field_map`. This way we can ignore the leadership stage column, as well as any arbitrary fields we add for reporting

cecilia-donnelly commented on streetcrm/views.py:
```diff
@@ -800,11 +812,12 @@ class EventParticipantsImport(FileImportMixin, generic.View):
         "Phone number": "primary_phone",
         "Email address": "email",
         "Address": "participant_street_address",
-        "Major action": "major_action"
```
Was this ever supposed to be included?

pjsier replied:
Nope, that was a typo

cecilia-donnelly commented:
The file I was testing with has the following contents: 
```
Exported on 2017-11-15,Search terms: search_for = participant,,,,,
,,,,,,
ID,Participant name,Phone number,Email address,Address,Institution,Leadership level
Fergus Anderson,17737032121,fergus@example.com,,St. Mary's of the Sea,
Mariella Sanchez,13125554321,msanchez@example.com,,St. Mary's of the Sea,
```

I skipped the "ID" column because I didn't want to accidentally cause ID collisions - maybe that was the problem?

pjsier commented:
@cecilia-donnelly ah so there were a few issues. The first is that I was only looking for entirely blank rows on the import, as opposed to one that has comma delimiters but is empty, so I added a check for that. Thanks for catching that!

The second is that the `id` key in the dictionary wasn't getting removed if it was blank, which meant that the queries would look for models with an empty ID and always return nothing. That's fixed as well.

The last thing is that right now the example you tried to import has the name in the `ID` column, but I tested it with a slight modification and it's working now:

```
Exported on 2017-11-15,Search terms: search_for = participant,,,,,
,,,,,,
ID,Participant name,Phone number,Email address,Address,Institution,Leadership level
,Fergus Anderson,17737032121,fergus@example.com,,St. Mary's of the Sea,
,Mariella Sanchez,13125554321,msanchez@example.com,,St. Mary's of the Sea,
```

cecilia-donnelly commented:
Hey @pjsier, thanks for the edited import file.  Now, if I try to import a file with a new person, with a new ID, like so: 
```
Exported on 2017-11-22,Search terms: search_for = participant,,,,,
,,,,,,
ID,Participant name,Phone number,Email address,Address,Institution,Leadership level
17,Chris Butler,,,,,
```
The new participant is created with the given ID, but the name is not filled in.  That is, an empty "participant 17" is created.  I presume this is a fairly unlikely scenario.  Is there a way to make it work, or make it error?  Maybe we need an error that says "This ID does not correspond to an existing participant.  Please re-import with an empty ID column." or something like that?

Similarly, if I import a file with the same ID but different information, nothing is changed.  I think this is because of our decision not to do updates with imports, right?

In any case, I'm getting a bit into the weeds.  I think we can merge this and ask them to test it.  Thanks!
